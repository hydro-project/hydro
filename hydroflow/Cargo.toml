[package]
name = "hydroflow"
version = "0.1.0"
edition = "2021"

[features]
default = [ "async", "macros" ]
async = [ "futures" ]
macros = [ "hydroflow_macro", "hydroflow_datalog" ]

[[example]]
name = "echoserver"

[[example]]
name = "echo_serde_json"

[[example]]
name = "chat"

[[example]]
name = "graph_reachability"

[[example]]
name = "graph_unreachability"

[[example]]
name = "three_clique"

[[example]]
name = "two_pc"

[[example]]
name = "deadlock_detector"

[[example]]
name = "kvs"

[dependencies]
bincode = "1.3"
byteorder = "1.4.3"
bytes = "1.1.0"
futures = { version = "0.3", optional = true }
hydroflow_datalog = { optional = true, path = "../hydroflow_datalog" }
hydroflow_lang = { path = "../hydroflow_lang" }
hydroflow_macro = { optional = true, path = "../hydroflow_macro" }
pin-project-lite = "0.2.9"
pusherator = { path = "../pusherator" }
ref-cast = "1.0"
regex = "1"
sealed = "0.4"
serde = { version = "1", features = [ "derive" ] }
serde_json = "1"
slotmap = "1.0"
static_assertions = "1.1.0"
tokio = { version = "1.16", features = [ "full" ] }
tokio-stream = { version = "0.1.10", features = [ "io-util" ] }
tokio-util = { version = "0.7.4", features = [ "net", "codec" ] }
variadics = { path = "../variadics" }

[dev-dependencies]
chrono = { version = "0.4.20", features = [ "serde" ], default-features = false }
clap = { version = "3.0.14", features = [ "derive" ] }
colored = "2.0"
criterion = { version = "0.3", features = [ "async_tokio" ] }
futures = { version = "0.3" }
itertools = "0.10.3"
rand = "0.8.4"
textnonce = "1.0.0"
time = "0.3"
trybuild = "1.0.71"
hdrhistogram = "7"
zipf = "7.0.0"
core_affinity = "0.5.10"
