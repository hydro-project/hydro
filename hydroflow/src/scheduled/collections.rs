pub struct Single<T>(pub T);

pub struct Iter<I: Iterator>(pub I);
