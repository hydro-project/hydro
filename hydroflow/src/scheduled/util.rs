use std::cell::RefCell;
use std::rc::Rc;

/**
 * Creates a once channel where the [SendOnce] end can set a value which the [Once] end will receive and store.
 */
pub fn once<T>() -> (SendOnce<T>, Once<T>) {
    let channel = Rc::new(RefCell::new(None));
    let sender = SendOnce {
        channel: channel.clone(),
    };
    let once = Once::Unresolved { channel };
    (sender, once)
}

/**
 * The sending half of a once channel.
 */
pub struct SendOnce<T> {
    channel: Rc<RefCell<Option<T>>>,
}
impl<T> SendOnce<T> {
    pub fn send(self, item: T) {
        let old_item = self.channel.borrow_mut().replace(item);
        assert!(old_item.is_none());
    }
}

/**
 * The receiving half of a once channel.
 */
pub enum Once<T> {
    Resolved(T),
    Unresolved { channel: Rc<RefCell<Option<T>>> },
    Errored,
}
impl<T> Once<T> {
    pub fn get(&mut self) -> &mut T {
        match self {
            Self::Resolved(item) => return item,
            Self::Errored => panic!("Initialization previously failed."),
            _ => {}
        }

        *self = match std::mem::replace(self, Self::Errored) {
            Self::Unresolved { channel } => {
                let item = Rc::try_unwrap(channel)
                    .ok()
                    .expect("Initialization failed: channel is still live.")
                    .into_inner()
                    .expect("Initialization failed: item never sent.");
                Self::Resolved(item)
            }
            v => v,
        };
        self.get()
    }
}
