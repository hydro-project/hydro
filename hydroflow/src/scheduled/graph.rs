use std::any::Any;
use std::collections::VecDeque;
use std::marker::PhantomData;
use std::sync::mpsc::{self, Receiver, RecvError, SyncSender};

use ref_cast::RefCast;

use super::context::Context;
use super::handoff::{Handoff, HandoffMeta, RecvPortList, SendPortList};
use super::port::{InputPort, OutputPort, RecvCtx, SendCtx};
use super::reactor::Reactor;
use super::state::StateHandle;
#[cfg(feature = "variadic_generics")]
use super::subgraph::Subgraph;
use super::{HandoffId, StateId, SubgraphId};

/**
 * A Hydroflow graph. Owns, schedules, and runs the compiled subgraphs.
 */
pub struct Hydroflow {
    subgraphs: Vec<SubgraphData>,
    handoffs: Vec<HandoffData>,

    states: Vec<StateData>,

    // TODO(mingwei): separate scheduler into its own struct/trait?
    ready_queue: VecDeque<SubgraphId>,
    event_queue_send: SyncSender<SubgraphId>, // TODO(mingwei) remove this, to prevent hanging.
    event_queue_recv: Receiver<SubgraphId>,
}
impl Default for Hydroflow {
    fn default() -> Self {
        let (subgraphs, handoffs, states, ready_queue) = Default::default();
        let (event_queue_send, event_queue_recv) = mpsc::sync_channel(8_000);
        Self {
            subgraphs,
            handoffs,
            states,
            ready_queue,
            event_queue_send,
            event_queue_recv,
        }
    }
}
impl Hydroflow {
    /**
     * Create an new empty Dataflow graph.
     */
    pub fn new() -> Self {
        Default::default()
    }

    /**
     * Returns a reactor for externally scheduling subgraphs, possibly from another thread.
     */
    pub fn reactor(&self) -> Reactor {
        Reactor::new(self.event_queue_send.clone())
    }

    fn enqueue_jobs(&mut self) {
        for sg in self.event_queue_recv.try_iter() {
            if !self.subgraphs[sg].scheduled {
                self.ready_queue.push_back(sg);
                self.subgraphs[sg].scheduled = true;
            }
        }
    }

    /**
     * Runs the dataflow until no more work is currently available.
     */
    pub fn tick(&mut self) {
        // Add any external jobs to ready queue.
        self.enqueue_jobs();

        while let Some(sg_id) = self.ready_queue.pop_front() {
            self.subgraphs[sg_id].scheduled = false;
            let sg_data = self.subgraphs.get_mut(sg_id).unwrap(/* TODO(mingwei) */);
            let context = Context {
                subgraph_id: sg_id,
                handoffs: &mut self.handoffs,
                states: &mut self.states,
                event_queue_send: &mut self.event_queue_send,
            };
            sg_data.subgraph.run(context);
            for &handoff_id in sg_data.succs.iter() {
                let handoff = self.handoffs.get(handoff_id).unwrap(/* TODO(mingwei) */);
                for &succ_id in handoff.succs.iter() {
                    if self.ready_queue.contains(&succ_id) {
                        // TODO(mingwei): Slow? O(N)
                        continue;
                    }
                    if !handoff.handoff.is_bottom() {
                        self.ready_queue.push_back(succ_id);
                    }
                }
            }

            self.enqueue_jobs();
        }
    }

    /**
     * Run the dataflow graph, blocking until completion.
     */
    pub fn run(&mut self) -> Result<!, RecvError> {
        loop {
            self.tick();
            self.poll_events()?;
        }
    }

    /**
     * Run the dataflow graph to completion asynchronously.
     */
    pub async fn run_async(&mut self) -> Result<!, RecvError> {
        loop {
            self.tick();
            self.poll_events()?;
            tokio::task::yield_now().await;
            // TODO(mingwei): this busy-spins when other tasks are not running.
        }
    }

    /**
     * Block and wait for an external event.
     */
    pub fn poll_events(&mut self) -> Result<(), RecvError> {
        self.ready_queue.extend(self.event_queue_recv.try_iter());
        Ok(())
    }

    /**
     * TODO(mingwei): Hack to re-enqueue all subgraphs.
     */
    pub fn wake_all(&mut self) {
        self.ready_queue.clear();
        self.ready_queue.extend(0..self.subgraphs.len());
    }

    /// Adds a new compiled subgraph with the specified inputs and outputs.
    ///
    /// See [TODO] for how to specify inputs and outputs.
    pub fn add_subgraph<R, W, F>(
        &mut self,
        recv_ports: R,
        send_ports: W,
        mut subgraph: F,
    ) -> SubgraphId
    where
        R: 'static + RecvPortList,
        W: 'static + SendPortList,
        F: 'static + FnMut(&Context<'_>, R::Ctx<'_>, W::Ctx<'_>),
    {
        let sg_id = self.subgraphs.len();

        let (mut subgraph_preds, mut subgraph_succs) = Default::default();
        recv_ports.set_graph_meta(&mut *self.handoffs, None, Some(sg_id), &mut subgraph_preds);
        send_ports.set_graph_meta(&mut *self.handoffs, Some(sg_id), None, &mut subgraph_succs);

        let subgraph = move |context: Context<'_>| {
            let recv = recv_ports.make_ctx(context.handoffs);
            let send = send_ports.make_ctx(context.handoffs);
            (subgraph)(&context, recv, send);
        };
        self.subgraphs
            .push(SubgraphData::new(subgraph, subgraph_preds, subgraph_succs));
        self.ready_queue.push_back(sg_id);

        sg_id
    }

    /// Adds a new compiled subraph with a variable number of inputs and outputs of the same respective handoff types.
    pub fn add_subgraph_n_m<R, W, F>(
        &mut self,
        recv_ports: Vec<OutputPort<R>>,
        send_ports: Vec<InputPort<W>>,
        mut subgraph: F,
    ) -> SubgraphId
    where
        R: 'static + Handoff,
        W: 'static + Handoff,
        F: 'static + FnMut(&Context<'_>, &[&RecvCtx<R>], &[&SendCtx<W>]),
    {
        let sg_id = self.subgraphs.len();

        let subgraph_preds = recv_ports.iter().map(|port| port.handoff_id).collect();
        let subgraph_succs = send_ports.iter().map(|port| port.handoff_id).collect();

        for recv_port in recv_ports.iter() {
            self.handoffs[recv_port.handoff_id].succs.push(sg_id);
        }
        for send_port in send_ports.iter() {
            self.handoffs[send_port.handoff_id].preds.push(sg_id);
        }

        let subgraph = move |context: Context<'_>| {
            let recvs: Vec<&RecvCtx<R>> = recv_ports
                .iter()
                .map(|hid| hid.handoff_id)
                .map(|hid| context.handoffs.get(hid).unwrap())
                .map(|h_data| {
                    h_data
                        .handoff
                        .any_ref()
                        .downcast_ref()
                        .expect("Attempted to cast handoff to wrong type.")
                })
                .map(RefCast::ref_cast)
                .collect();

            let sends: Vec<&SendCtx<W>> = send_ports
                .iter()
                .map(|hid| hid.handoff_id)
                .map(|hid| context.handoffs.get(hid).unwrap())
                .map(|h_data| {
                    h_data
                        .handoff
                        .any_ref()
                        .downcast_ref()
                        .expect("Attempted to cast handoff to wrong type.")
                })
                .map(RefCast::ref_cast)
                .collect();

            (subgraph)(&context, &recvs, &sends)
        };
        self.subgraphs
            .push(SubgraphData::new(subgraph, subgraph_preds, subgraph_succs));
        self.ready_queue.push_back(sg_id);

        sg_id
    }

    pub fn make_handoff<H>(&mut self) -> (InputPort<H>, OutputPort<H>)
    where
        H: 'static + Handoff,
    {
        let handoff_id: HandoffId = self.handoffs.len();

        // Create and insert handoff.
        let handoff = H::default();
        self.handoffs.push(HandoffData::new(handoff));

        // Make ports.
        let input_port = InputPort {
            handoff_id,
            _marker: PhantomData,
        };
        let output_port = OutputPort {
            handoff_id,
            _marker: PhantomData,
        };
        (input_port, output_port)
    }

    pub fn add_state<T>(&mut self, state: T) -> StateHandle<T>
    where
        T: Any,
    {
        let state_id: StateId = self.states.len();

        let state_data = StateData {
            state: Box::new(state),
        };
        self.states.push(state_data);

        StateHandle {
            state_id,
            _phantom: PhantomData,
        }
    }
}

/**
 * A handoff and its input and output [SubgraphId]s.
 *
 * NOT PART OF PUBLIC API.
 */
pub struct HandoffData {
    pub(crate) handoff: Box<dyn HandoffMeta>,
    pub(crate) preds: Vec<SubgraphId>,
    pub(crate) succs: Vec<SubgraphId>,
}
impl std::fmt::Debug for HandoffData {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        f.debug_struct("HandoffData")
            .field("preds", &self.preds)
            .field("succs", &self.succs)
            .finish_non_exhaustive()
    }
}
impl HandoffData {
    pub fn new(handoff: impl 'static + HandoffMeta) -> Self {
        let (preds, succs) = Default::default();
        Self {
            handoff: Box::new(handoff),
            preds,
            succs,
        }
    }
}

/**
 * A subgraph along with its predecessor and successor [SubgraphId]s.
 * Used internally by the [Hydroflow] struct to represent the dataflow graph structure.
 */
struct SubgraphData {
    subgraph: Box<dyn Subgraph>,
    #[allow(dead_code)]
    preds: Vec<HandoffId>,
    succs: Vec<HandoffId>,
    scheduled: bool,
}
impl SubgraphData {
    pub fn new(
        subgraph: impl 'static + Subgraph,
        preds: Vec<HandoffId>,
        succs: Vec<HandoffId>,
    ) -> Self {
        Self {
            subgraph: Box::new(subgraph),
            preds,
            succs,
            scheduled: true,
        }
    }
}

pub struct StateData {
    pub state: Box<dyn Any>,
}
