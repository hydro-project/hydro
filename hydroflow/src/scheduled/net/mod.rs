use std::collections::VecDeque;
use std::pin::Pin;

use futures::{Sink, StreamExt};
use tokio::net::{TcpListener, TcpStream};
use tokio_util::codec::{FramedRead, FramedWrite, LengthDelimitedCodec};

use super::{
    ctx::{InputPort, OutputPort, RecvCtx},
    handoff::VecHandoff,
    Hydroflow,
};

const MESSAGE_DATA: u8 = 0;

const TYPE_LEN: usize = 1;
const ADDRESS_LEN: usize = 4;

#[derive(Clone, Debug)]
pub enum Message {
    Data { address: u32, batch: bytes::Bytes },
}

impl Message {
    fn encode(&self, v: &mut Vec<u8>) {
        match self {
            Message::Data { address, batch } => {
                v.push(MESSAGE_DATA);
                v.extend((*address as u32).to_ne_bytes());
                v.extend(batch);
            }
        }
    }

    fn decode(v: &bytes::Bytes) -> Self {
        match v[0] {
            MESSAGE_DATA => {
                let address =
                    u32::from_ne_bytes(v[TYPE_LEN..(TYPE_LEN + ADDRESS_LEN)].try_into().unwrap());
                let batch = v.slice((TYPE_LEN + ADDRESS_LEN)..);
                Message::Data { address, batch }
            }
            _ => panic!("unhandled"),
        }
    }
}

impl Hydroflow {
    fn add_tcp_stream(
        &mut self,
        stream: TcpStream,
    ) -> (
        InputPort<VecHandoff<Message>>,
        OutputPort<VecHandoff<Message>>,
    ) {
        let (reader, writer) = stream.into_split();
        let reader = FramedRead::new(reader, LengthDelimitedCodec::new());
        let reader_port = self.add_input_from_stream::<_, VecHandoff<_>, _>(
            reader.map(|buf| Some(<Message>::decode(&buf.unwrap().into()))),
        );
        let mut writer = FramedWrite::new(writer, LengthDelimitedCodec::new());

        let mut message_queue = VecDeque::new();

        let writer_port: InputPort<VecHandoff<Message>> =
            self.add_sink(move |ctx, recv: &RecvCtx<VecHandoff<Message>>| {
                let waker = ctx.waker();
                let mut cx = std::task::Context::from_waker(&waker);

                // TODO(mingwei): queue may grow unbounded? Subtle rate matching concern.
                // TODO(mingwei): put into state system.
                message_queue.extend(recv.take_inner().into_iter());
                while !message_queue.is_empty() {
                    if let std::task::Poll::Ready(Ok(())) =
                        Pin::new(&mut writer).poll_ready(&mut cx)
                    {
                        let v = message_queue.pop_front().unwrap();
                        let mut buf = Vec::new();
                        v.encode(&mut buf);

                        Pin::new(&mut writer).start_send(buf.into()).unwrap();
                    }
                }
                let _ = Pin::new(&mut writer).poll_flush(&mut cx);
            });
        (writer_port, reader_port)
    }

    // Connects to the specified address, returning an input and output port
    // allowing communication on it.
    pub async fn connect(
        &mut self,
        addr: &str,
    ) -> (
        InputPort<VecHandoff<Message>>,
        OutputPort<VecHandoff<Message>>,
    ) {
        let stream = TcpStream::connect(addr).await.unwrap();
        self.add_tcp_stream(stream)
    }

    // Waits for a single connection on the specified unix port, returning an input
    // and output port allowing communication on it.
    pub async fn bind_one(
        &mut self,
        port: usize,
    ) -> (
        InputPort<VecHandoff<Message>>,
        OutputPort<VecHandoff<Message>>,
    ) {
        let stream = TcpListener::bind(format!("localhost:{}", port))
            .await
            .unwrap();
        let (stream, _) = stream.accept().await.unwrap();
        self.add_tcp_stream(stream)
    }
}
