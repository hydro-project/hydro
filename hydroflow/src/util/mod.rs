//! Helper utilities for the Hydroflow surface syntax.

mod udp;
pub use udp::*;
mod tcp;
pub use tcp::*;

use std::net::SocketAddr;
use std::task::{Context, Poll};

use bincode;
use futures::Stream;
use serde::{Deserialize, Serialize};

pub fn unbounded_channel<T>() -> (
    tokio::sync::mpsc::UnboundedSender<T>,
    tokio_stream::wrappers::UnboundedReceiverStream<T>,
) {
    let (send, recv) = tokio::sync::mpsc::unbounded_channel();
    let recv = tokio_stream::wrappers::UnboundedReceiverStream::new(recv);
    (send, recv)
}

/// Collects the immediately available items into a `FromIterator` collection.
///
/// This consumes the stream, use [`futures::StreamExt::by_ref()`] if you want
/// to retain ownership of your stream.
pub fn recv_into<C, S>(stream: S) -> C
where
    C: FromIterator<S::Item>,
    S: Stream,
{
    let mut stream = Box::pin(stream);
    std::iter::from_fn(|| {
        match stream
            .as_mut()
            .poll_next(&mut Context::from_waker(futures::task::noop_waker_ref()))
        {
            Poll::Ready(opt) => opt,
            Poll::Pending => None,
        }
    })
    .collect()
}

pub fn serialize_msg<T>(msg: T) -> bytes::Bytes
where
    T: Serialize + for<'a> Deserialize<'a> + Clone,
{
    bytes::Bytes::from(bincode::serialize(&msg).unwrap())
}

pub fn deserialize_simple<T>(msg: bytes::BytesMut) -> T
where
    T: Serialize + for<'a> Deserialize<'a> + Clone,
{
    bincode::deserialize(&msg).unwrap()
}

pub fn deserialize_msg<T>(
    msg: Result<(bytes::BytesMut, SocketAddr), tokio_util::codec::LinesCodecError>,
) -> T
where
    T: Serialize + for<'a> Deserialize<'a> + Clone,
{
    bincode::deserialize(&(msg.unwrap().0)).unwrap()
}

pub fn ipv4_resolve(addr: &str) -> Result<SocketAddr, std::io::Error> {
    use std::net::ToSocketAddrs;
    let mut addrs = addr.to_socket_addrs()?;
    let result = addrs.find(|addr| addr.is_ipv4());
    match result {
        Some(addr) => Ok(addr),
        None => Err(std::io::Error::new(
            std::io::ErrorKind::Other,
            "Unable to resolve IPv4 address",
        )),
    }
}

pub async fn bind_udp_bytes(addr: SocketAddr) -> (UdpSink, UdpStream) {
    let socket = tokio::net::UdpSocket::bind(addr).await.unwrap();
    udp_bytes(socket)
}

pub async fn bind_udp_lines(addr: SocketAddr) -> (UdpLinesSink, UdpLinesStream) {
    let socket = tokio::net::UdpSocket::bind(addr).await.unwrap();
    udp_lines(socket)
}
