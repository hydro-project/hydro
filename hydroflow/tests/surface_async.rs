//! Surface syntax tests of asynchrony and networking.

use std::collections::HashSet;
use std::error::Error;
use std::net::{Ipv4Addr, SocketAddr};
use std::time::Duration;

use bytes::Bytes;
use hydroflow::scheduled::graph::Hydroflow;
use hydroflow::util::{collect_ready, tcp_lines};
use hydroflow::{hydroflow_syntax, rassert, rassert_eq};
use tokio::net::{TcpListener, TcpStream, UdpSocket};
use tokio::task::LocalSet;
use tokio_util::codec::{BytesCodec, FramedWrite, LinesCodec};

#[tokio::test]
pub async fn test_echo_udp() -> Result<(), Box<dyn Error>> {
    let local = LocalSet::new();

    let server_socket = UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0)).await?;
    let server_addr = server_socket.local_addr()?;
    let server_addr: SocketAddr = (Ipv4Addr::LOCALHOST, server_addr.port()).into();

    // Server:
    let serv = local.spawn_local(async {
        let socket = server_socket;
        let (udp_send, udp_recv, _) = hydroflow::util::udp_lines(socket);
        println!("Server live!");

        let (seen_send, seen_recv) = hydroflow::util::unbounded_channel();

        let mut df: Hydroflow = hydroflow_syntax! {
            recv = source_stream(udp_recv)
                -> map(|r| r.unwrap())
                -> tee();
            // Echo
            recv[0] -> dest_sink(udp_send);
            // Testing
            recv[1] -> map(|(s, _addr)| s) -> for_each(|s| seen_send.send(s).unwrap());
        };

        tokio::select! {
            _ = df.run_async() => (),
            _ = tokio::time::sleep(Duration::from_secs(1)) => (),
        };

        let seen: HashSet<_> = collect_ready(seen_recv);
        rassert_eq!(4, seen.len())?;
        rassert!(seen.contains("Hello"))?;
        rassert!(seen.contains("World"))?;
        rassert!(seen.contains("Raise"))?;
        rassert!(seen.contains("Count"))?;

        Ok(()) as Result<(), Box<dyn Error>>
    });

    // Client A:
    let client_a = local.spawn_local(async move {
        let socket = UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0))
            .await
            .unwrap();
        let (send_udp, recv_udp, _) = hydroflow::util::udp_lines(socket);

        let (seen_send, seen_recv) = hydroflow::util::unbounded_channel();

        let mut df = hydroflow_syntax! {
            recv = source_stream(recv_udp)
                -> map(|r| r.unwrap())
                -> tee();
            recv[0] -> for_each(|x| println!("client A recv: {:?}", x));
            recv[1] -> map(|(s, _addr)| s) -> for_each(|s| seen_send.send(s).unwrap());

            // Sending
            source_iter([ "Hello", "World" ]) -> map(|s| (s.to_owned(), server_addr)) -> dest_sink(send_udp);
        };

        tokio::select! {
            _ = df.run_async() => (),
            _ = tokio::time::sleep(Duration::from_secs(1)) => (),
        };

        let seen: Vec<_> = collect_ready(seen_recv);
        rassert_eq!(&["Hello".to_owned(), "World".to_owned()], &*seen)?;

        Ok(()) as Result<(), Box<dyn Error>>
    });

    // Client B:
    let client_b = local.spawn_local(async move {
        let socket = UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0))
            .await
            .unwrap();
        let (send_udp, recv_udp, _) = hydroflow::util::udp_lines(socket);

        let (seen_send, seen_recv) = hydroflow::util::unbounded_channel();

        let mut df = hydroflow_syntax! {
            recv = source_stream(recv_udp)
                -> map(|r| r.unwrap())
                -> tee();
            recv[0] -> for_each(|x| println!("client B recv: {:?}", x));
            recv[1] -> map(|(s, _addr)| s) -> for_each(|s| seen_send.send(s).unwrap());

            // Sending
            source_iter([ "Raise", "Count" ]) -> map(|s| (s.to_owned(), server_addr)) -> dest_sink(send_udp);
        };

        tokio::select! {
            _ = df.run_async() => (),
            _ = tokio::time::sleep(Duration::from_secs(1)) => (),
        };

        let seen: Vec<_> = collect_ready(seen_recv);
        rassert_eq!(&["Raise".to_owned(), "Count".to_owned()], &*seen)?;

        Ok(()) as Result<(), Box<dyn Error>>
    });

    local.await;
    serv.await??;
    client_a.await??;
    client_b.await??;

    Ok(())
}

#[tokio::test]
pub async fn test_echo_tcp() -> Result<(), Box<dyn Error>> {
    let local = LocalSet::new();

    // Port 0 -> picks any available port.
    let listener = TcpListener::bind((std::net::Ipv4Addr::LOCALHOST, 0)).await?;
    let addr = listener.local_addr()?;

    // Server:
    let serv = local.spawn_local(async move {
        let (server_stream, _) = listener.accept().await?;
        let (server_send, server_recv) = tcp_lines(server_stream);

        println!("Server accepted connection!");

        let (seen_send, seen_recv) = hydroflow::util::unbounded_channel();

        let mut df: Hydroflow = hydroflow_syntax! {
            rev = source_stream(server_recv)
                -> map(|x| x.unwrap())
                -> tee();
            rev[0] -> dest_sink(server_send);
            rev[1] -> for_each(|s| seen_send.send(s).unwrap());
        };

        tokio::time::timeout(Duration::from_secs(1), df.run_async())
            .await
            .expect_err("Expected time out");

        let seen: Vec<_> = collect_ready(seen_recv);
        rassert_eq!(&["Hello".to_owned(), "World".to_owned()], &*seen)?;

        Ok(()) as Result<(), Box<dyn Error>>
    });

    // Client:
    let client = local.spawn_local(async move {
        let client_stream = TcpStream::connect(addr).await?;
        let (client_send, client_recv) = tcp_lines(client_stream);

        println!("Client connected!");

        let (seen_send, seen_recv) = hydroflow::util::unbounded_channel();

        let mut df = hydroflow_syntax! {
            recv = source_stream(client_recv)
                -> map(|x| x.unwrap())
                -> tee();

            recv[0] -> for_each(|s| println!("echo {}", s));
            recv[1] -> for_each(|s| seen_send.send(s).unwrap());

            source_iter([
                "Hello",
                "World",
            ]) -> dest_sink(client_send);
        };

        println!("Client running!");

        tokio::time::timeout(Duration::from_secs(1), df.run_async())
            .await
            .expect_err("Expected time out");

        let seen: Vec<_> = collect_ready(seen_recv);
        rassert_eq!(&["Hello".to_owned(), "World".to_owned()], &*seen)?;

        Ok(()) as Result<(), Box<dyn Error>>
    });

    local.await;
    serv.await??;
    client.await??;

    Ok(())
}

#[tokio::test]
pub async fn test_echo() {
    // An edge in the input data = a pair of `usize` vertex IDs.
    let (lines_send, lines_recv) = hydroflow::util::unbounded_channel::<String>();

    // LinesCodec separates each line from `lines_recv` with `\n`.
    let stdout_lines = FramedWrite::new(tokio::io::stdout(), LinesCodec::new());

    let mut df: Hydroflow = hydroflow_syntax! {
        source_stream(lines_recv) -> dest_sink(stdout_lines);
    };

    println!(
        "{}",
        df.serde_graph()
            .expect("No graph found, maybe failed to parse.")
            .to_mermaid()
    );
    df.run_available();

    lines_send.send("Hello".to_owned()).unwrap();
    lines_send.send("World".to_owned()).unwrap();
    df.run_available();

    lines_send.send("Hello".to_owned()).unwrap();
    lines_send.send("World".to_owned()).unwrap();
    df.run_available();

    // Allow background thread to catch up.
    tokio::time::sleep(Duration::from_secs(1)).await;
}

#[tokio::test]
pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
    const MAX: usize = 20;

    let (mut send, recv) = hydroflow::futures::channel::mpsc::channel::<usize>(5);
    send.try_send(0).unwrap();

    let (seen_send, seen_recv) = hydroflow::util::unbounded_channel();

    let mut df = hydroflow_syntax! {
        recv = source_stream(recv) -> tee();
        recv[0] -> map(|x| x + 1)
            -> filter(|&x| x < MAX)
            -> dest_sink(send);
        recv[1] -> for_each(|x| seen_send.send(x).unwrap());
    };

    tokio::select! {
        _ = df.run_async() => (),
        _ = tokio::time::sleep(Duration::from_secs(1)) => (),
    };

    let seen: Vec<_> = collect_ready(seen_recv);
    rassert_eq!(&std::array::from_fn::<_, MAX, _>(|i| i), &*seen)?;

    Ok(())
}

#[tokio::test(flavor = "current_thread")]
async fn asynctest_dest_sink_bounded_channel() {
    // In this example we use a _bounded_ channel for our `Sink`. This is for demonstration only,
    // instead you should use [`hydroflow::util::unbounded_channel`]. A bounded channel results in
    // `Hydroflow` buffering items internally instead of within the channel.
    let (send, recv) = tokio::sync::mpsc::channel::<usize>(5);
    let send = tokio_util::sync::PollSender::new(send);
    let mut recv = tokio_stream::wrappers::ReceiverStream::new(recv);

    let mut flow = hydroflow_syntax! {
        source_iter(0..10) -> dest_sink(send);
    };
    tokio::time::timeout(std::time::Duration::from_secs(1), flow.run_async())
        .await
        .expect_err("Expected time out");

    // Only 5 elemts received due to buffer size
    let out: Vec<_> = collect_ready(&mut recv);
    assert_eq!(&[0, 1, 2, 3, 4], &*out);
}

#[tokio::test(flavor = "current_thread")]
async fn asynctest_dest_sink_duplex() {
    use bytes::Bytes;
    use tokio::io::AsyncReadExt;
    use tokio_util::codec;

    // Like a channel, but for a stream of bytes instead of discrete objects.
    let (asyncwrite, mut asyncread) = tokio::io::duplex(256);
    // Now instead handle discrete byte lists by length-encoding them.
    let mut sink = codec::LengthDelimitedCodec::builder()
        // Use 1 byte len field (max 255) so we don't have to worry about endianness.
        .length_field_length(1)
        .new_write(asyncwrite);

    let mut flow = hydroflow_syntax! {
        source_iter([
            Bytes::from_static(b"hello"),
            Bytes::from_static(b"world"),
        ]) -> dest_sink(&mut sink);
    };
    tokio::time::timeout(std::time::Duration::from_secs(1), flow.run_async())
        .await
        .expect_err("Expected time out");

    let mut buf = Vec::<u8>::new();
    asyncread.read_buf(&mut buf).await.unwrap();
    // `\x05` is length prefix of "5".
    assert_eq!(b"\x05hello\x05world", &*buf);
}

#[tokio::test(flavor = "current_thread")]
async fn asynctest_dest_asyncwrite_duplex() {
    use tokio::io::AsyncReadExt;

    // Like a channel, but for a stream of bytes instead of discrete objects.
    // This could be an output file, network port, stdout, etc.
    let (asyncwrite, mut asyncread) = tokio::io::duplex(256);
    let sink = FramedWrite::new(asyncwrite, BytesCodec::new());

    let mut flow = hydroflow_syntax! {
        source_iter([
            Bytes::from_static("hello".as_bytes()),
            Bytes::from_static("world".as_bytes()),
        ]) -> dest_sink(sink);
    };
    tokio::time::timeout(std::time::Duration::from_secs(1), flow.run_async())
        .await
        .expect_err("Expected time out");

    let mut buf = Vec::<u8>::new();
    asyncread.read_buf(&mut buf).await.unwrap();
    // `\x05` is length prefix of "5".
    assert_eq!(b"helloworld", &*buf);
}

#[tokio::test(flavor = "current_thread")]
async fn asynctest_source_stream() {
    LocalSet::new()
        .run_until(async {
            let (a_send, a_recv) = hydroflow::util::unbounded_channel::<usize>();
            let (b_send, b_recv) = hydroflow::util::unbounded_channel::<usize>();

            tokio::task::spawn_local(async move {
                let mut flow = hydroflow_syntax! {
                    source_stream(a_recv) -> for_each(|x| { b_send.send(x).unwrap(); });
                };
                flow.run_async().await.unwrap();
            });
            tokio::task::spawn_local(async move {
                let mut flow = hydroflow_syntax! {
                    source_stream(b_recv) -> for_each(|x| println!("{}", x));
                };
                flow.run_async().await.unwrap();
            });

            a_send.send(1).unwrap();
            a_send.send(2).unwrap();
            a_send.send(3).unwrap();

            tokio::task::yield_now().await;
        })
        .await;
}
