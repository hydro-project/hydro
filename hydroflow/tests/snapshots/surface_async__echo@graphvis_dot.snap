---
source: hydroflow/tests/surface_async.rs
expression: df.serde_graph().unwrap().to_dot()
---
digraph {
    subgraph "cluster n1v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_1v1\nstratum 0"
        n1v1 [label="(n1v1) source_stream(lines_recv)", fontname=Monaco, shape=invhouse, style = filled, color = "#0022ff", fontcolor = "#ffffff"]
        n2v1 [label="(n2v1) dest_sink(stdout_lines)", fontname=Monaco, shape=house, style = filled, color = "#ffff00"]
        n1v1 -> n2v1
    }
}

